realm/realm-java
# Kotlin Extension Library

Closes #4092
Closes #4701

This PR contains a new dependency `realm-android-kotlin-extensions` that is automatically included in all projects including Kotlin. The primary purpose to provide extension functions for methods so they become more Kotlin friendly.

TODO 
- [x] Figure out how and what kind of docs we should distribute. KDoc or Javadoc? KDoc seems to work fine since IntelliJ apparently uses the sources for the inline docs. CMD + F1 is showing correct and jumping to the definition also has the docs as expected.
- [x] Fix wrong classes included in KDoc: Currently not possible https://youtrack.jetbrains.com/issue/KT-11333
- [x] Test upload to OJO
- [x] Test upload to Bintray
- [x] Correctly report unit test failures
- [x] KDoc for all methods
- [x] Add unit tests for all extension methods
- [ ] Update website docs

Usage of these extensions can be seen here: https://github.com/realm/realm-java/pull/4684/files#diff-a9bdfddc109df70caf521cb459b12d15

---- 

OUTDATED

This PR is just a proof of concept for extending our current Kotlin support and is not intended to be merged as it is, but more like a starting point for a discussion. With Kotlin now being an official language for Android, we should explore the options that don't require rewriting everything in Kotlin ;)

This PR outlines an approach where we create a new `io.realm:realm-android-kotlin-extension` artifact that our Gradle plugin can automatically include if Kotlin is used, just like we detect when to use `kapt` today.

What do you think @realm/java ?

The question then remains what to put into such a plugin. Low hanging fruits would be:
- Add extension functions for all methods taking `Class` and allow `KClass` instead.
- Consider adding support for referencing field names directly for type safe queries. Will not work for chained queries and require a dependency for `kotlin-reflect` :/
- Add support for Kotlins null/non-null types. Cannot be part of the plugin, but must live in the annotation processor. It just require us detecting `org.jetbrains.annotations.NotNull` as a substitute for `@Required`.
- Other ideas?
